packages/automerge-repo/src/storage/StorageSubsystem.ts

```diff
@@ -1,46 +1,148 @@
+    /** Namespace to prevent collisions with other users of the storage subsystem. */
+    namespace: string,
+
+    /** Key to load. Typically a UUID or other unique identifier, but could be any string. */
+    key: string,
 
-  constructor(storageAdapter: StorageAdapter) {
-    this.#storageAdapter = storageAdapter
+    /** Data to save, as a binary blob. */
+    data: Uint8Array
+  ): Promise<void> {
+    const storageKey = [namespace, key] as StorageKey
+    await this.storageAdapter.save(storageKey, data)
   }
 
+  /** Removes a value from storage. */
+  async remove(
+    /** Namespace to prevent collisions with other users of the storage subsystem. */
+    namespace: string,
+
+    /** Key to remove. Typically a UUID or other unique identifier, but could be any string. */
+    key: string
+  ): Promise<void> {
+    const storageKey = [namespace, key] as StorageKey
+    await this.storageAdapter.remove(storageKey)
+  }
+
+  // AUTOMERGE DOCUMENT STORAGE
+
+  /**
+   * Loads the Automerge document with the given ID from storage.
+   */
+  async loadDoc<T>(documentId: DocumentId): Promise<A.Doc<T> | null> {
+    // Load all the chunks for this document
+    const chunks = await this.storageAdapter.loadRange([documentId])
+    const binaries = []
+    const chunkInfos: ChunkInfo[] = []
+
+    for (const chunk of chunks) {
+      const chunkType = chunkTypeFromKey(chunk.key)
+      if (chunkType == null) continue
+      chunkInfos.push({
+        key: chunk.key,
+        type: chunkType,
+        size: chunk.data.length,
+      })
+      binaries.push(chunk.data)
+    }
+    this.#chunkInfos.set(documentId, chunkInfos)
+
+    // Merge the chunks into a single binary
+    const binary = mergeArrays(binaries)
+    if (binary.length === 0) return null
+
+    // Load into an Automerge document
+    const newDoc = A.loadIncremental(A.init(), binary) as A.Doc<T>
+
+    // Record the latest heads for the document
+    this.#storedHeads.set(documentId, A.getHeads(newDoc))
+
+    return newDoc
+  }
+
+  /**
+   * Saves the provided Automerge document to storage.
+   *
+   * @remarks
+   * Under the hood this makes incremental saves until the incremental size is greater than the
+   * snapshot size, at which point the document is compacted into a single snapshot.
+   */
+  async saveDoc(documentId: DocumentId, doc: A.Doc<unknown>): Promise<void> {
+    // Don't bother saving if the document hasn't changed
+    if (!this.#shouldSave(documentId, doc)) return
+
+    const sourceChunks = this.#chunkInfos.get(documentId) ?? []
+    if (this.#shouldCompact(sourceChunks)) {
+      void this.#saveTotal(documentId, doc, sourceChunks)
+    } else {
+      void this.#saveIncremental(documentId, doc)
+    }
+    this.#storedHeads.set(documentId, A.getHeads(doc))
+  }
+
+  /**
+   * Removes the Automerge document with the given ID from storage
+   */
+  async removeDoc(documentId: DocumentId) {
+    void this.storageAdapter.removeRange([documentId, "snapshot"])
```

Comment:
is it worth doing an indexedDB migration to clean this up and stop squatting on the whole documentId namespace?

Reply:
Oh I really don't think so. Maybe it would have been cleaner to do it that way from the beginning but if there's not a problem I would just leave it. As long as nobody but the storage subsystem talks directly to the storage adapters it should be fine.